docs(specs): Labels and Selectors for Plugin Enablement

neelayu opened the pull request:
## Summary
Provides a spec for Plugin enable and disablement using labels and selectors.

## Checklist
<!-- Mandatory
Please confirm the following by replacing the space with an "x" between the []:
-->

- [x] No AI generated code was used in this PR

## Related issues
<!-- Mandatory
All PRs should resolve an issue, if one does not exist, please open one.
-->

resolves NA
related #16704

Hipska commented on docs/specs/tsd-010-labels-and-selectors.md:
```diff
@@ -0,0 +1,79 @@
+
+## Configuration Options
+
+###### Labels
+Labels are key-value pairs provided at runtime and attached to the Telegraf instance.
```
```suggestion
###### Labels

Labels are key-value pairs provided at runtime and attached to the Telegraf instance.
```

Hipska commented on docs/specs/tsd-010-labels-and-selectors.md:
````diff
@@ -0,0 +1,79 @@
+
+Each plugin definition in the configuration file can optionally include a selectors field.
+
+Example:
+```toml
````
```suggestion
Example:

```toml
```

Hipska commented on docs/specs/tsd-010-labels-and-selectors.md:
````diff
@@ -0,0 +1,79 @@
+[[inputs.cpu]]
+  selectors = ["app=payments,region=us-east", "env=prod"]
+```
+###### Boolean behaviour
+- Multiple selectors are treated as **OR**
+
+- Within a single selector string, multiple conditions separated by commas are treated as **AND**
````
```suggestion
###### Boolean behaviour

- Multiple selectors are treated as **OR**
- Within a single selector string, multiple conditions separated by commas are treated as **AND**
```

neelayu replied:
Done

Hipska reviewed (changes requested):
The content of the spec looks good to me. Only some markup changes requested.

Please also increase/decrease headings by 1 per step. (not 3)

neelayu commented on docs/specs/tsd-010-labels-and-selectors.md:
```diff
@@ -0,0 +1,79 @@
+| `app=web, region=us-east, env=prod` | `["app=web,env=prod", "region=eu-*"]` | First selector matches (`app=web` **AND** `env=prod`) | Selected |
+| `app=worker, region=us-central`     | `["region=us-*"]`                     | Wildcard match on `region=us-central`                 | Selected |
+| `app=worker, region=us-central`     | `["region=eu-*"]`                     | Wildcard mismatch                                     | Skipped  |
+| `app=api`                           | `["app=web", "app=api"]`              | Second selector matches (`app=api`)                   | Selected |
```
Agreed. made the change

neelayu commented on docs/specs/tsd-010-labels-and-selectors.md:
````diff
@@ -0,0 +1,79 @@
+
+At Telegraf startup, users can provide labels via a CLI flag:
+
+```bash
````
Done

neelayu commented on docs/specs/tsd-010-labels-and-selectors.md:
````diff
@@ -0,0 +1,79 @@
+```
+
+###### Selectors
+A selector is a string describing one or more required key-value matches.
````
Done

Hipska commented on docs/specs/tsd-010-labels-and-selectors.md:
```diff
@@ -0,0 +1,83 @@
+
+### Matching Examples
+
+| Labels Provided                     | Plugin Selectors                      | Matching Behavior                                     | Result     |
+| :---------------------------------- | :------------------------------------ | :---------------------------------------------------- | :--------- |
+| `app=web`, `region=us-east`           | `["app=web"]`                         | `app=web` matches exactly                             | Selected |
+| `app=web`, `region=us-east`           | `["app=api"]`                         | `app=api` does not match                              | Skipped  |
+| `app=web`, `region=us-east`           | `["app=web,region=us-east"]`          | Both `app=web` **AND** `region=us-east` match         | Selected |
+| `app=web`, `region=us-east`           | `["app=web,region=us-west"]`          | `region=us-west` does **not** match                   | Skipped  |
+| `app=web`, `region=us-east`, `env=prod` | `["app=web,env=prod", "region=eu-*"]` | First selector matches (`app=web` **AND** `env=prod`) | Selected |
+| `app=worker`, `region=us-central`     | `["region=us-*"]`                     | Wildcard match on `region=us-central`                 | Selected |
+| `app=worker`, `region=us-central`     | `["region=eu-*"]`                     | Wildcard mismatch                                     | Skipped  |
+| `app=api`                           | `["app=web", "app=api"]`              | Second selector matches (`app=api`)                   | Selected |
```
Please apply my suggestion if you want to keep the table aligned..
```suggestion
| Labels Provided                         | Plugin Selectors                      | Matching Behavior                                     | Result   |
| :-------------------------------------- | :------------------------------------ | :---------------------------------------------------- | :------- |
| `app=web`, `region=us-east`             | `["app=web"]`                         | `app=web` matches exactly                             | Selected |
| `app=web`, `region=us-east`             | `["app=api"]`                         | `app=api` does not match                              | Skipped  |
| `app=web`, `region=us-east`             | `["app=web,region=us-east"]`          | Both `app=web` **AND** `region=us-east` match         | Selected |
| `app=web`, `region=us-east`             | `["app=web,region=us-west"]`          | `region=us-west` does **not** match                   | Skipped  |
| `app=web`, `region=us-east`, `env=prod` | `["app=web,env=prod", "region=eu-*"]` | First selector matches (`app=web` **AND** `env=prod`) | Selected |
| `app=worker`, `region=us-central`       | `["region=us-*"]`                     | Wildcard match on `region=us-central`                 | Selected |
| `app=worker`, `region=us-central`       | `["region=eu-*"]`                     | Wildcard mismatch                                     | Skipped  |
| `app=api`                               | `["app=web", "app=api"]`              | Second selector matches (`app=api`)                   | Selected |

neelayu replied:
Done.

neelayu commented on docs/specs/tsd-010-labels-and-selectors.md:
```diff
@@ -0,0 +1,83 @@
+### Use Case
+
+- Deploy multiple Telegraf instances fetching configurations from a centralized source.
+- Activate only the plugins relevant to each instance based on labels and selectors.
```
Done

neelayu commented:
Hi @srebhan 
thanks for reviewing. I have addressed some of the comments- I think you got labels and selectors mixed up. Labels will be simply a list, and selectors will be following logical operations depending on the input.
Accordingly I have made the changes. Let me know.

neelayu commented:
@srebhan 

> Furthermore, I do have a question: With the current description how do you select
(app=web AND region=us-east) OR (app=backend AND region=eu-central)  with the --label flags? I think you need to support lists also on the command-line flag to do this. 

This would mean we pass
```toml
selectors = ["app=web,region=us-east", "app=backend,region=eu-central"]
```

To match it against the labels, we require a list of labels and it is documented in this spec. We would pass one or more `--label` flags 

```console
--label="app=web" --label="region=us-east" --label="another=value"
```

This would match for the first selector. This case is present in the 5th example of the table.

neelayu commented on docs/specs/tsd-010-labels-and-selectors.md:
```diff
@@ -0,0 +1,83 @@
+
+[Issue #1317](https://github.com/influxdata/telegraf/issues/1317)
+
+[Issue #10543](https://github.com/influxdata/telegraf/issues/10543)
```
probably got overwritten. thanks

neelayu commented on docs/specs/tsd-010-labels-and-selectors.md:
```diff
@@ -0,0 +1,90 @@
+
+## Keywords
+
+configuration Management, dynamic plugin selection
```
done

srebhan commented:
@neelayu I think

> This would match for the first selector. This case is present in the 5th example of the table.

is not what I mean. All of your examples show an `OR` combination, i.e. as soon as at least on `label` provided on the command line matches the plugin is selected.

What I mean is that if you allow the selectors to be individual entries like

```toml
[[inputs.foo]]
  selectors = ["app=web", "region=us-east"]

[[inputs.foo]]
  selectors = ["app=web", "region=us-west"]

[[inputs.foo]]
  selectors = ["app=web", "region=eu-central"]

[[inputs.foo]]
  selectors = ["app=backend", "region=us-east"]

[[inputs.foo]]
  selectors = ["app=backend", "region=us-west"]

[[inputs.foo]]
  selectors = ["app=backend", "region=eu-central"]
```

You have more rich ways of selecting groups like `--label "app=web"` or `--label="region=us-*"` at command line and if you allow for `--label="app=web,region=us-east"` you can enforce matching a subset. That's much easier than doing it the other way around, isn't it?

neelayu commented:
@srebhan I see your point.

Can we get @Hipska's thoughts as well? @Hipska seemed okay with this spec. Just curious if the labels and selector terms are making sense or if something’s unclear.

Hipska commented:
I'm just another guy from the community, just like You..

I was okay with the original spec indeed. However, I also agree with what @srebhan said. For me, both ways would work.

**Note:** I don't actually need this functionality as I generate the config for each system. Hint: You can use a generic `telegraf.conf` and have system specific configs in `telegraf.d/`.

neelayu commented:
Thanks @Hipska 
I just wanted to clarify whether the terminology might be causing any confusion. In the Kubernetes world, the typical phrasing is "Select all objects that match the labels." This is because labels are considered atomic key-value pairs, meaning they don't support wildcards. 


@srebhan 
Let's consider the following case-
there are two operators responsible for running telegraf instances.
They will run these instances with say the following commands
```console
Telegraf-01
---
telegraf --config-directory configs/ --label="env=prod" --label="cpu=true" --label="instance=us-west"
```

```console
Telegraf-02
telegraf --config-directory configs/ --label="env=prod" --label="mem=true"
```

Now there could be many individuals who write their configs accordingly
```toml
[[inputs.cpu]]
# The first selector does not match with any label, but second one mandates env=prod AND cpu=true
# Hence Telegraf-01 will "select" this plugin
selector = ["instance=eu-*", "env=prod,cpu=true"]
```

```toml
[[inputs.mem]]
# The selector matches with Telegraf-02 and hence it will be selected on that instance
selector = ["env=prod,mem=true"]
```

Now the downside could be if there is the following config

```toml
[[inputs.mem]]
selector = ["instance=us-*"]
```
This will match Telegraf-01 and hence will be selected.
Essentially, if there is a disjoint between operators and config writers, then such cases can arise even when we invert the terminolgy. But I feel operators will run these instances with appropriate labels only to ensure config writers know where the plugin is to be selected.

If the confusion is wrt to labels and selector terminology only, then I can definitely rework the definitions.

The solution would be to ensure we thoroughly document the behaviour

Operator Responsibility:
Operators must ensure labels are specific and accurate to avoid potential overlaps or unintended plugin activation.

Config Writer Responsibility:
Config writers need to be mindful of selectors and avoid overly broad patterns that could match unintended instances.

neelayu commented:
That is true. But this scenario works if you want those plugins to be enabled. For disabling it, you will either comment the block out or remove it completely. Of course, one would need a watch to ensure this happens without restarting the telegraf.

As mentioned in the original issue, I am looking for a more robust way to handle it.

Hipska commented:
For disabling it, it is just matter of not including that specific config file. Which is indeed the same effect as commenting or removing the block. Seems like the most robust as it can get, don't you think?

Hipska commented:
Did you actually see my response? It is about adding/removing the `--config` option from the command line. Like you would also add or remove `--label` from it. It doesn't require/need touching the files at all..

neelayu commented:
Not really. I would change the selector in the plugin to ensure it doesn't match the label and hence get dropped by telegraf without having to stop the running instance of telegraf

neelayu commented:
Yes that is correct. Let me work through it.

neelayu commented:
@srebhan I have made the changes, can you review it again? thanks!

srebhan commented on docs/specs/tsd-010-labels-and-selectors.md:
```diff
@@ -0,0 +1,130 @@
+
+Telegraf must implement a new optional `labels` configuration setting. This
+setting must be available in all input, output, aggregator and processor plugins.
+The `labels` configuration setting must accept a map where each entry is a single
+key-value pair in the form `<key>=<value>`.
```
```suggestion
The `labels` configuration setting must accept a map where each entry is a single
key-value pair.
```

srebhan reviewed (commented):
Awesome @neelayu! Just a few more wording improvements (at least IMO ;-))...

srebhan reviewed (approved):
Thanks a lot @neelayu!

neelayu commented:
Thanks @srebhan and @Hipska for your valuable suggestions!
